fabric-transfer-api-v1/src/main/java/net/fabricmc/fabric/impl/transfer/fluid/FluidKeyImpl.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.transfer.fluid;
+
+import java.util.Objects;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.fluid.Fluid;
+import net.minecraft.fluid.Fluids;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.network.PacketByteBuf;
+import net.minecraft.util.Identifier;
+import net.minecraft.util.registry.Registry;
+
+import net.fabricmc.fabric.api.transfer.v1.fluid.FluidKey;
+
+public class FluidKeyImpl implements FluidKey {
+	public static FluidKey of(Fluid fluid, @Nullable CompoundTag tag) {
+		Objects.requireNonNull(fluid, "Fluid may not be null.");
+
+		if (!fluid.isStill(fluid.getDefaultState())) {
+			throw new IllegalArgumentException("Fluid may not be flowing.");
+		}
+
+		// TODO cache
+		return new FluidKeyImpl(fluid, tag);
```

Comment:
FluidKeys without data should be cached, I don't think you need to cache those with tags

Reply:
Should I cache them inside `Fluid` directly? That's what I was planning on doing for `ItemKey`s. ^^

Reply:
yea, or even just implement FluidKey on Fluid itself

Reply:
I think storing it inside `Fluid` is cleaner.

Reply:
Done.